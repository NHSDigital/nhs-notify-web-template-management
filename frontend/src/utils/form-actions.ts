'use server';

import { getSessionServer } from '@utils/amplify-utils';
import {
  CreateUpdateTemplate,
  isTemplateDtoValid,
  RoutingConfig,
  RoutingConfigStatusActive,
  routingConfigurationApiClient,
  TemplateDto,
  ValidatedTemplateDto,
} from 'nhs-notify-backend-client';
import { logger } from 'nhs-notify-web-template-management-utils/logger';
import { templateApiClient } from 'nhs-notify-backend-client/src/template-api-client';
import { sortAscByCreatedAt } from './sort';

export async function createTemplate(
  template: CreateUpdateTemplate
): Promise<TemplateDto> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.createTemplate(
    template,
    accessToken
  );

  if (error) {
    logger.error('Failed to create template', { error });
    throw new Error('Failed to create new template');
  }

  return data;
}

export async function uploadLetterTemplate(
  template: CreateUpdateTemplate,
  pdf: File,
  csv: File
) {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.uploadLetterTemplate(
    template,
    accessToken,
    pdf,
    csv?.size > 0 ? csv : undefined
  );

  if (error) {
    logger.error('Failed to create letter template', { error });
    throw new Error('Failed to create new letter template');
  }

  return data;
}

export async function saveTemplate(
  template: Extract<TemplateDto, { templateType: 'EMAIL' | 'SMS' | 'NHS_APP' }>
): Promise<TemplateDto> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.updateTemplate(
    template.id,
    template,
    accessToken
  );

  if (error) {
    logger.error('Failed to save template', { error });
    throw new Error('Failed to save template data');
  }

  return data;
}

export async function setTemplateToSubmitted(
  templateId: string
): Promise<TemplateDto> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.submitTemplate(
    templateId,
    accessToken
  );

  if (error) {
    logger.error('Failed to save template', { error });
    throw new Error('Failed to save template data');
  }

  return data;
}

export async function setTemplateToDeleted(templateId: string): Promise<void> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { error } = await templateApiClient.deleteTemplate(
    templateId,
    accessToken
  );

  if (error) {
    logger.error('Failed to save template', { error });
    throw new Error('Failed to save template data');
  }
}

export async function requestTemplateProof(
  templateId: string
): Promise<TemplateDto> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.requestProof(
    templateId,
    accessToken
  );

  if (error) {
    logger.error('Failed to request proof', { error });
    throw new Error('Failed to request proof');
  }

  return data;
}

export async function getTemplate(
  templateId: string
): Promise<TemplateDto | undefined> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.getTemplate(
    templateId,
    accessToken
  );

  if (error) {
    logger.error('Failed to get template', { error });
  }

  return data;
}

export async function getTemplates(): Promise<TemplateDto[]> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.listTemplates(accessToken);

  if (error) {
    logger.error('Failed to get templates', { error });
    return [];
  }

  const sortedData = data
    .map((template) => isTemplateDtoValid(template))
    .filter(
      (template): template is ValidatedTemplateDto => template !== undefined
    );

  return sortAscByCreatedAt(sortedData);
}

export async function getRoutingConfigs(): Promise<RoutingConfig[]> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await routingConfigurationApiClient.list(accessToken);

  if (error) {
    logger.error('Failed to get routing configuration', {
      error: error,
    });
    return [];
  }

  return sortAscByCreatedAt(data);
}

export async function countRoutingConfigs(
  status: RoutingConfigStatusActive
): Promise<number> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await routingConfigurationApiClient.count(
    accessToken,
    status
  );

  if (error) {
    logger.error(`Failed to count routing configuration for ${status}`, {
      error,
    });
    return 0;
  }

  return data.count;
}
