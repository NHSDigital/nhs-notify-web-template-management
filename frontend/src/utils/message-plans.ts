import {
  type RoutingConfig,
  $RoutingConfig,
  type RoutingConfigStatusActive,
} from 'nhs-notify-backend-client';
import { routingConfigurationApiClient } from 'nhs-notify-backend-client/src/routing-config-api-client';
import { logger } from 'nhs-notify-web-template-management-utils/logger';
import { getSessionServer } from './amplify-utils';
import { sortAscByUpdatedAt } from './sort';

export async function getRoutingConfigs(): Promise<RoutingConfig[]> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await routingConfigurationApiClient.list(accessToken);

  if (error) {
    logger.error('Failed to get routing configuration', error);
    return [];
  }

  const valid = data.filter((d) => {
    const { error: validationError, success } = $RoutingConfig.safeParse(d);

    if (!success) {
      logger.error('Listed invalid routing configuration', validationError);
    }

    return success;
  });

  return sortAscByUpdatedAt(valid);
}

export async function countRoutingConfigs(
  status: RoutingConfigStatusActive
): Promise<number> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await routingConfigurationApiClient.count(
    accessToken,
    status
  );

  if (error) {
    logger.error(`Failed to count routing configuration for ${status}`, {
      error,
    });
    return 0;
  }

  return data.count;
}
