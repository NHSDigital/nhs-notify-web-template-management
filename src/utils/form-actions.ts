'use server';

import { getAmplifyBackendClient } from '@utils/amplify-utils';
import { DbOperationError } from '@domain/errors';
import { Template, Draft } from './types';
import { logger } from './logger';

export async function createTemplate(
  template: Draft<Template>
): Promise<Template> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.create(template);

  if (errors || !data) {
    logger.error('Failed to create template', errors);
    throw new Error('Failed to create new template');
  }
  return data;
}

export async function saveTemplate(template: Template): Promise<Template> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.update(template);

  if (errors) {
    logger.error('Failed to save template', errors);
    throw new Error('Failed to save template data');
  }

  if (!data) {
    throw new DbOperationError({
      message:
        'Template in unknown state. No errors reported but entity returned as falsy',
      operation: 'update',
    });
  }
  return data;
}

export async function getTemplate(
  templateId: string
): Promise<Template | undefined> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.get({
      id: templateId,
    });

  if (errors) {
    logger.error('Failed to get template', errors);
  }

  if (!data) {
    logger.warn(`Failed to retrieve template for ID ${templateId}`);
    return undefined;
  }

  return data;
}

export async function sendEmail(
  templateId: string,
  templateName: string,
  templateMessage: string,
  templateSubjectLine: string | null
) {
  const res = await getAmplifyBackendClient().queries.sendEmail({
    recipientEmail: 'england.test.cm@nhs.net',
    templateId,
    templateName,
    templateMessage,
    templateSubjectLine,
  });

  if (res.errors) {
    logger.error({
      description: 'Error sending email',
      res,
    });
  }
}

export async function getTemplates(): Promise<Template[] | []> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.list();

  if (errors) {
    logger.error('Failed to get templates', errors);
  }

  if (!data) {
    logger.warn(`Failed to retrieve templates`);
    return [];
  }

  return data;
}
