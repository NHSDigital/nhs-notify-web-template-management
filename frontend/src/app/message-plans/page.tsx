'use server';

import content from '@content/content';
import { MessagePlans } from '@molecules/MessagePlans/MessagePlans';
import { Metadata } from 'next';
import { countRoutingConfigs, getRoutingConfigs } from '@utils/form-actions';

export async function generateMetadata(): Promise<Metadata> {
  return {
    title: content.pages.messagePlansPage.pageTitle,
  };
}

const MessagePlansPage = async () => {
  const [routingConfigurations, draftCount, completedCount] = await Promise.all(
    [
      getRoutingConfigs(),
      countRoutingConfigs('DRAFT'),
      countRoutingConfigs('COMPLETED'),
    ]
  );

  const messagePlans = routingConfigurations.map((plan) => ({
    name: plan.name,
    id: plan.id,
    lastUpdated: plan.updatedAt,
    status: plan.status,
  }));

  const draft = messagePlans.filter((r) => r.status === 'DRAFT');

  const completed = messagePlans.filter((r) => r.status === 'COMPLETED');

  return (
    <MessagePlans
      draft={{
        plans: draft,
        count: draftCount,
      }}
      production={{
        plans: completed,
        count: completedCount,
      }}
    />
  );
};

export default MessagePlansPage;
