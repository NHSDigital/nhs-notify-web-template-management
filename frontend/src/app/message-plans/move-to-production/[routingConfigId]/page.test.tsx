import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { mock } from 'jest-mock-extended';
import { redirect, RedirectType } from 'next/navigation';
import { RoutingConfig } from 'nhs-notify-backend-client';
import { verifyFormCsrfToken } from '@utils/csrf-utils';
import { submitRoutingConfig } from '@utils/form-actions';
import { getRoutingConfig } from '@utils/message-plans';
import Page, { metadata } from './page';
import { randomUUID } from 'node:crypto';

jest.mock('next/navigation');
jest.mock('@utils/csrf-utils');
jest.mock('@utils/message-plans');
jest.mock('@utils/form-actions');
jest.mocked(verifyFormCsrfToken).mockResolvedValue(true);

const routingConfig = mock<RoutingConfig>({
  name: 'My Routing Config',
});

beforeEach(() => {
  jest.clearAllMocks();
});

test('metadata', () => {
  expect(metadata).toEqual({
    title: 'Move message plan to production - NHS Notify',
  });
});

test('matches snapshot', async () => {
  jest.mocked(getRoutingConfig).mockResolvedValueOnce(routingConfig);

  const page = await Page({
    params: Promise.resolve({ routingConfigId: 'routing-config-id' }),
  });

  const container = render(page);

  expect(container.asFragment()).toMatchSnapshot();
});

test('redirects if routing config is not found from path parameter', async () => {
  await Page({
    params: Promise.resolve({ routingConfigId: 'routing-config-id' }),
  });

  expect(getRoutingConfig).toHaveBeenCalledWith('routing-config-id');

  expect(redirect).toHaveBeenCalledWith(
    '/message-plans/invalid',
    RedirectType.replace
  );
});

test('redirects if form is submitted with invalid routing config id', async () => {
  jest.mocked(getRoutingConfig).mockResolvedValueOnce(routingConfig);

  const user = await userEvent.setup();

  const page = await Page({
    params: Promise.resolve({ routingConfigId: 'routing-config-id' }),
  });

  render(page);

  await user.click(screen.getByTestId('submit-button'));

  expect(redirect).toHaveBeenCalledWith(
    '/message-plans/invalid',
    RedirectType.push
  );

  expect(submitRoutingConfig).not.toHaveBeenCalled();
});

test('submits the message plan and redirects when form is submitted', async () => {
  const routingConfigId = randomUUID();

  jest.mocked(getRoutingConfig).mockResolvedValueOnce(routingConfig);

  const user = await userEvent.setup();

  const page = await Page({
    params: Promise.resolve({ routingConfigId }),
  });

  render(page);

  await user.click(screen.getByTestId('submit-button'));

  expect(redirect).toHaveBeenCalledWith('/message-plans', RedirectType.push);

  expect(submitRoutingConfig).toHaveBeenCalledWith(routingConfigId);
});
