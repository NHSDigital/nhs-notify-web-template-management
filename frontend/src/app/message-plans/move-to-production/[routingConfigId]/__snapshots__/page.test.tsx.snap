// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="nhsuk-main-wrapper"
    id="maincontent"
    role="main"
  >
    <div
      class="nhsuk-grid-row"
    >
      <div
        class="nhsuk-grid-column-two-thirds"
      >
        <h1
          class="nhsuk-heading-xl"
        >
          Move message plan to production
        </h1>
        <table
          class="nhsuk-u-margin-bottom-6 table"
        >
          <tbody
            class="nhsuk-table__body"
          >
            <tr
              class="nhsuk-table__row"
            >
              <th
                class="nhsuk-table__cell"
              >
                Name
              </th>
              <td
                class="nhsuk-table__cell"
              >
                My Routing Config
              </td>
              <td
                class="nhsuk-table__cell"
              >
                <a
                  href="/message-plans/preview-message-plan/routing-config-id"
                >
                  Preview
                </a>
              </td>
            </tr>
          </tbody>
        </table>
        <p>
          Moving message plans from draft to production means that they are ready to send.
        </p>
        <p>
          Messages will only be sent to recipients when you make a request with
          <a
            href="https://digital.nhs.uk/developer/api-catalogue/nhs-notify"
            rel="noopener noreferrer"
            target="_blank"
          >
            NHS Notify API
          </a>
           or
          <a
            href="https://digital.nhs.uk/developer/api-catalogue/nhs-notify-mesh"
            rel="noopener noreferrer"
            target="_blank"
          >
            NHS Notify MESH
          </a>
          .
        </p>
        <h2
          class="nhsuk-heading-m"
          id="before-you-move-message-plans-to-production"
        >
          Before you move message plans to production
        </h2>
        <p>
          Any templates used in these message plans will be locked.
        </p>
        <p>
          Make sure:
        </p>
        <ul
          class="nhsuk-list nhsuk-list--bullet"
        >
          <li>
            the relevant stakeholders in your team have approved your templates and message plan
          </li>
          <li>
            your templates have no errors
          </li>
        </ul>
        <div
          class="nhsuk-warning-callout"
        >
          <h3
            class="nhsuk-warning-callout__label"
          >
            Important
            <span
              class="nhsuk-u-visually-hidden"
            >
              :
            </span>
          </h3>
          <p>
            You cannot edit anything that is in production.
          </p>
          <p>
            If you need to edit your templates or message plans, you can copy and replace them.
          </p>
        </div>
        <form
          action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        >
          <input
            name="form-id"
            readonly=""
            type="hidden"
            value="move-message-plan-to-production"
          />
          <input
            name="csrf_token"
            readonly=""
            type="hidden"
            value="no_token"
          />
          <input
            name="routingConfigId"
            readonly=""
            type="hidden"
            value="routing-config-id"
          />
          <div
            class="nhsuk-form-group"
          >
            <button
              aria-disabled="false"
              class="nhsuk-button nhsuk-button--warning"
              data-testid="submit-button"
              type="submit"
            >
              Move to production
            </button>
            <a
              aria-disabled="false"
              class="nhsuk-button nhsuk-button--secondary nhsuk-u-margin-left-3"
              draggable="false"
              href="/templates/message-plans"
              role="button"
            >
              Keep in draft
            </a>
          </div>
        </form>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
